### DISCUSSION TITLE
Make the $hookName variable in widget based modules available in templates by default

### DISCUSSION THREAD

**Discussion by dennispw** [PrestaShop/PrestaShop#33422] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

The widget functionality is great since it allows for any module that uses it to be placed, pretty much, anywhere. However, this also means that the markup might be "wrong" depending on the hook.

As an example, let's take the standard `ps_currencyselector` together with the default `classic` theme. It contains two sets of code, one dropdown alà Bootstrap and one standard `select` element. It is moved by JavaScript depending on the viewport width and the `prestashop.responsive` object.

Or what if we have a module that need's a `h2` instead of a `h4` depending on the position? Something just as simple as a different class?

If we had the `$hookName` variable available we could more easily split the code up for specific positions, remove any moving-of-elements with JavaScript and be rid of specific elements, with specific IDs for specific positions for specific themes.

Now, depending on the module it's quite easy to use overrides and add that variable in. However, if the module have `private` or `protected` function(s) it can quite quickly become a complete rewrite of the module which obivously is bad for many other reasons. Future update(s) is just one of them.

### Describe the solution you'd like

Make the `$hookName` variable available by default for widget modules.

Let's again use the `ps_currencyselector` as an example:

Now instead of having BOTH a Bootstrap flavoured dropdown AND a native hidden `select` element we could do something like this:

```
{if $hookName == 'displayFoo'}
  {* Dropdown markup *}
{elseif $hookName == 'displayBar'}
  {* Select markup *}
{/if}
```

You could perhaps do some variable juggling:
```
{if $hookName == 'displayFoo'}
  {assign var="somethingCool" value=true}
  {assign var="someClasses" value="text-primary text-uppercase"}
{else}
  {assign var="somethingCool" value=false}
  {assign var="someClasses" value="text-secondary"}
{/if}
```

### Alternatives you've considered

Overriding modules is pretty much the only way to do this right now if I'm not missing anything.

**Comment by kpodemski** [comment]
Thank you, @dennispw. I like your idea and agree with it.

Would you like to have `$hookName` automatically available in the view or have this variable added in the native modules?

**Comment by dennispw** [comment]
> Thank you, @dennispw. I like your idea and agree with it.
> 
> Would you like to have `$hookName` automatically available in the view or have this variable added in the native modules?

If it's possible to have it available by default in every view that uses the widget functionality that would be best way forward. 

The other option of updating _each and every native module that uses it_ is, I guess, another solution, although in my opinion a very poor one. It will most likely take a *long* time and I'm guessing a large chunk of the modules will be, for lack of a better term here, forgotten about.

**Comment by marwachelly** [comment]
Hello @dennispw 

Thank you for you report. Could you please check this issue https://github.com/PrestaShop/PrestaShop/issues/20164. Isn't the same?

Thanks,

**Comment by dennispw** [comment]
> Hello @dennispw
> 
> Thank you for you report. Could you please check this issue #20164. Isn't the same?
> 
> Thanks,

I'm not sure. If I understand that issue correctly the desired solution is _only_ be able to use different `.tpl` files?

My proposed solution is to still use "one" `.tpl` file. In it you could possibly do something like:
```
{if $hookName == 'displayFoo'}
  {include file='module:somemodule/_partials/some_template.tpl'}
{elseif $hookName == 'displayBar'}
  {include file='module:somemodule/_partials/another_template.tpl'}
{/if}
```

But it would also allow you to do more simple tasks such as switching out an element, add/remove classes or whatever it might be. It's not _solely_ for using different `.tpl` files entirely.

Obviously it should still be possible to use specific `.tpl` files for a specific `hook` if needed. Such as:
```
public function hookDisplaySomething()
{
  return $this->fetch('module:somemodule/views/templates/hook/specific_template.tpl');
}
```

But as soon as the actual widget is used, make the `$hookName` variable available by default.

As another example of how useful it can be is in the `ps_linklist` module. Since it can be useful for it to be in the footer, header, perhaps a right/left column or on the product page at the same time. We could easily adjust it to be a dropdown, a list, an inline list and more, something like:
```
{if $linkList.hook === 'displayProductAdditionalInfo'}
  {* Specific for the product page *}
{elseif $linkList.hook === 'displayNav1'}
  {* Dropdown markup *}
{elseif $linkList.hook === 'displayLeftColumn'}
  {assign var="showListTitle" value=false}
{/if}
```

**Comment by kpodemski** [comment]
Hello @marwachelly 

I think current issue provides a solution for the one you mentioned, and it is better described 👍

**Comment by marwachelly** [comment]
> Hello @marwachelly
> 
> I think current issue provides a solution for the one you mentioned, and it is better described +1

Hello @kpodemski 

Since this issue is duplicated to  #20164 it should be closed as duplicated and I suggest to update the description of the first one. What do you think?

Many thanks,

**Comment by dennispw** [comment]
@kpodemski @marwachelly 

I've asked the author of the other issue to check if _this_ issue is in any way related to the other issue. As I said, unfortunately I don't fully understand the other one.

Update: https://github.com/PrestaShop/PrestaShop/issues/20164#issuecomment-987871199 Seems these are separate issues even though some of the parts might be linked?

**Comment by marwachelly** [comment]
Hi @dennispw ,
Thank you for your suggestion. The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 
Thanks,

**Comment by micka-fdz** [comment]
Hi. I'm really interested in this feature.

@kpodemski any idea how can we do without BC break? Maybe in `getWidgetVariables()` function? I can make a PR.

**Comment by kpodemski** [comment]
@micka-fdz you can introduce the changes with BC Breaks since we're working on version 9.0 on the `develop` branch  👍🏻 but if you want to discuss detailed implementation feel free to ping me on the project's Slack 👍🏻

**Comment by micka-fdz** [comment]
@kpodemski I contacted you on FoP's Slack ✋

**Comment by kpodemski** [comment]
At first, I was in favor of doing this change, but I'm not sure how to approach it so the variable will be available only in the context of one template.

If anybody have an idea, let us know :)
